0027-tcp-load-balancer.md

```diff
@@ -0,0 +1,264 @@
+an appropriate server.
+The ServerGroup struct will house connection stats and addresser for each server.
+
+Connections will be incremented and decremented as connections are created and destroyed by clients.
+
+As such there will be an simple API that will increment or decrement the server's connection count.
+
+```rust
+struct ServerGroup
+{
+    id           : u32,
+    server_addrs : HashMap<u32, String>,
+    cxn_cntr     : HasnMap<u32, usize>,
+}
+
+impl ServerGroup
+{
+    fn find_min(&self) -> Option<u32>
+    {
+        // No clients yet
+        if self.cxn_cntr.len() == 0
+        {
+            for (id,addr) in self.server_addrs.iter()
+            {
+                return Some(id);
+            }
+        }
+
+        let mut min_conns = usize::MAX;
+        let mut min_id : Option<u32> = None;
+        for (id, num_conns) in self.cnx_cntr.iter()
+        {
+            if min > num_conns
+            {
+                min_conns = num_conns;
+                min_id = Some(id);
+            }
+        }
+
+        min_id
+    }
+}
+
+```
+
+#### Upstream Health Check
+
+The health checker will simply connect to an upstream server, send a ping and receive a pong message and then disconnect.
+
+It will continue to do this indefinitely every 30 seconds (hard coded for development speed).
+
+If the upstream server rejects the connection, does not reply to the ping  within 1 second or disconnects while the connection is active, the upstream server is marked unhealthy.
+
+If the upstream completes a full connect - ping - disconnect cycle, it is again marked healthy.
+
+```rust
+enum UpstreamState
+{
+    HEALTHY,
+    UNHEALTHY,
+}
+```
+
+```rust
+enum PingState
+{
+    IDLE,
+    PING_SENT(i64),
+    PONG_RECEIVED(i64),
+}
+```
+
+```rust
+struct HealthChecker
+{
+    up_stream       : TcpStream,
+    ping_state      : PingState,
+    UpstreamState   : UpstreamState,
+    last_update     : chrono::NaiveDateTime,
+}
+```
+
+```rust
+struct ServerGroup
+{
+    ...
+    server_health   : HashMap<u32, HealthChecker>,
+}
+```
+
+
+### Server
+
+The server will listen on port 8443 for client connections, using std::net::TcpListener to listen connections from clients.
+
+Each connection will be then authenticated by the rusttls server instance.
+
+Once authenticated, the authorisation process will occur as listed below.
+
+
+#### Security
+
+The rustls crate will be utilised as for all cryptogaphic functionality.
+- Since rustls has *NOT* had a security audit yet, this code should definitely *NOT* be used in production.
+- I found rustls to have a better API and docs (from my research) than native-tls.
+
+The x509-parser crate will be used for public key data extraction from client certificates provided to the server.
+
+TLS v1.3 with ECDHE will be used exclusively as it newer, has better performance on initial connection and gives perfect forward secrecy out of the box on all cipher suites.
```

Comment:
I don't think TLS 1.3 supports non-ephemeral key exchange anyway.

Reply:
https://www.davidwong.fr/tls13/#section-2

There is the possibility of using Pre-Shared Keys with or without (EC)DHE. But I think that is the only way.